### PR TITLE
Fix DateInput component deleted value should not produce error

### LINKED ISSUE #362: DateInput component: Erasing value in the input produces "Invalid date" message.
### Bug report

- __Package version(s)__: 1.3.1

#### Steps to reproduce

1. Click on Date Input and select date
2. Click mouse on DateInput input box and erase the text pressing backspace

#### Actual behavior

The text box displays "Invalid date" in red font:

screenshot 2016-12-13 20 04 57

#### Expected behavior

The text box displays Placeholder.

#### Reasoning

- User did not enter any date. they entered no date therefore "Invalid date" is not appropriate.
- User did not do anything wrong, they only unset the entered value intending not to provide any value.

### PR DESCRIPTION
**Pull request by AlexMarvelo**
#### Changes proposed in this pull request:

- on input blur if field is empty, no errors appear

### CODE REVIEW

**Comment by palantirtech** [comment]
Thanks for your interest in [palantir/blueprint](https://github.com/palantir/blueprint), @AlexMarvelo! Before we can accept your pull request, you need to sign our contributor license agreement - just visit https://cla.palantir.com/ and follow the instructions. Once you sign, I'll automatically update this pull request.

**Comment by romanr** [comment]
Tested in Chrome, IE, FF

File: packages/datetime/src/dateInput.tsx [hunk 1]
```diff
@@ -298,7 +298,9 @@ export class DateInput extends AbstractComponent<IDateInputProps, IDateInputStat
     private handleInputBlur = () => {
         const valueString = this.state.valueString;
         let value = moment(valueString, this.props.format);
-        if (valueString !== this.getDateString(this.state.value) && (!value.isValid() || !this.dateIsInRange(value))) {
+        if (valueString.length &&
```
Comment (adidahiya) [inline packages/datetime/src/dateInput.tsx]: please avoid implicit type coercion to booleans. instead use `valueString.length > 0`

Comment (adidahiya) [inline packages/datetime/src/dateInput.tsx]: nit: we prefer putting the binary operator on the next line for multiline expressions, like this:

```ts
if (valueString.length
    && valueString ...
```

(yes, this really should be enforced by our linter)

File: packages/datetime/src/dateInput.tsx [hunk 2]
```diff
@@ -314,7 +316,11 @@ export class DateInput extends AbstractComponent<IDateInputProps, IDateInputStat
                 Utils.safeInvoke(this.props.onChange, this.fromMomentToDate(value));
             }
         } else {
-            this.setState({ isInputFocused: false });
+            if (!valueString.length) {
```
Comment (adidahiya) [inline packages/datetime/src/dateInput.tsx]: same here, implicit type coercion

File: packages/datetime/src/dateInput.tsx [hunk 3]
```diff
@@ -314,7 +316,11 @@ export class DateInput extends AbstractComponent<IDateInputProps, IDateInputStat
...
+                this.setState({ isInputFocused: false, value: moment(null), valueString: null });
```
Comment (adidahiya) [inline packages/datetime/src/dateInput.tsx]: why does `valueString` need to be `null`; can it not be the empty string, or `undefined`?

Comment (AlexMarvelo) [inline packages/datetime/src/dateInput.tsx]: I tried to follow [your style](https://github.com/palantir/blueprint/blob/master/packages/datetime/src/dateInput.tsx#L304)

File: packages/datetime/src/dateInput.tsx (hunk 2 above)
Comment (AlexMarvelo) [inline packages/datetime/src/dateInput.tsx]: Fixed

File: packages/datetime/src/dateInput.tsx (hunk 1 above)
Comment (AlexMarvelo) [inline packages/datetime/src/dateInput.tsx]: Fixed

**Pull request merged**